README
Clutter version 1.002
=====================

Perl bindings to the Clutter OpenGL canvas.  Clutter aims to give a nice
and easy GObject based API for creating fast, heavily stylised applications,
such as media box UI, presentations, kiosk style applications, etc.

Clutter uses OpenGL for rendering.

The Clutter Perl bindings allow you to write canvas-based applications in
a Perlish and object-oriented way, freeing you from having to care about
casting and memory management of the original Clutter C API, and yet
remaining close to its spirit.

You can find out more about Clutter at http://www.clutter-project.org .

INSTALLATION
------------

To install this module type the following:

  perl Makefile.PL
  make
  make test
  make install

To avoid installing to a system directory you can change the installation
prefix at Makefile.PL time with

  perl Makefile.PL PREFIX=/some/other/place

This will install the module to the subdirectory lib/perl5 under the given
prefix.  If this is not already in Perl's include path, you'll need to tell
perl how to the to this library directory so you can use it.

DEPENDENCIES
------------

This module requires these other modules and libraries:

  perl >= 5.8.0
  Glib >= 1.180 (Perl module)
  Cairo >= 1.0.0 (Perl module)
  Pango >= 1.140 (Perl module)
  Clutter >= 1.0.0 (C library)

In order to build it from source, you'll also need:

  ExtUtils::Depends >= 0.2
  ExtUtils::PkgConfig >= 1.03

When building with support for Clutter-Cairo, the Cairo Perl module is
also required.

When building with support for Clutter-GStreamer, the GStreamer Perl
module is also required.

BUG REPORTS
-----------

Please report bugs to the Clutter product in OpenedHand's Bugzilla:

  http://bugzilla.gnome.org/enter_bug.cgi?product=Clutter

COPYRIGHT AND LICENSE
---------------------

Copyright (C) 2006  OpenedHand Ltd
Copyright (C) 2008  Intel Corporation

This library is free software; you can redistribute it and/or modify it under
the terms of the GNU Library General Public License as published by the Free
Software Foundation; either version 2.1 of the License, or (at your option)
any later version

This library is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU Library General Public License for more
details.

You should have received a copy of the GNU Library General Public License along
with this library; if not, write to the Free Software Foundation, Inc., 59
Temple Place - Suite 330, Boston, MA  02111-1307  USA.


NEWS
Clutter 1.001
=============
* Add Clutter::Cogl::Pango API bindings
* Fix a segmentation fault when parsing path nodes of type 'close'
* Add Clutter::Path::foreach() and Clutter::Path::to_cairo_path()
* Allow constructing Clutter::Path objects both by using the stringified
  SVG notation and by using lists of Clutter::PathNode objects
* Add more units to the test suite, covering Clutter::Path and Clutter::Event
* Fully document Clutter::Path
* Add an example of a Clutter::Container implementation
* Improve documentation

Clutter 1.000
=============
* First developers snapshot of the Clutter Perl bindings
* Depend on Clutter 1.0.0
* Drop the clutter-gtk, clutter-gst and clutter-cairo bindings
* Add more COGL bindings under Clutter::Cogl
* Port more examples from Clutter to Perl
* Provide more C API for bindings based on perl-Clutter
* Add ParamSpec bindings for colors and units

Clutter 0.820
=============
* Depend on Clutter 0.8.2
* Fix a typo in Clutter::Cogl->frustum()
* Add a Clutter::Cookbook POD

Clutter 0.802
=============
* Update the META.yml generation so that the contents are useful
  for CPAN and testers
* Add some initial wrappers for the COGL texture API; still
  experimental, will require more testing
* Fix the assignment of built-in alpha functions when creating
  a new Clutter::Alpha [Andrew Ruthven]
* Documentation fixes
* Better coverage for the test suite

Clutter 0.801
=============
* Compilation fixes required for compiling with Clutter installed
  from the 0.8.0 tarball
* Wrap Clutter::Gst::Audio::get_playbin()
* Wrap Clutter::Gst::VideoSink
* Enhance Clutter::Model and Clutter::Model::Iter bindings to
  allow pure-Perl implementations

Clutter 0.800
=============
* Require clutter-0.8 >= 0.8.0
* Resync with upstream API
* Wrap newly exposed COGL API as Clutter::Cogl
* Wrap newly added Gtk2::Clutter utility API
* Allow full implementation of Clutter::Container

Clutter 0.620
=============
* Require clutter-0.6 >= 0.6.2
* Require clutter-cairo-0.6 >= 0.6.1
* Require clutter-gtk-0.6 >= 0.6.1
* Add new API to Clutter::Texture::Cairo for sub-region updates
* Fix usage of glib-perl API
* Return a boolean value in Clutter::Entry->handle_key_event()
* Do no re-register GConnectFlags as a GType if it's already known

Clutter 0.600
=============
* Resync with upstream API

Clutter 0.500
=============
* Resync with upstream API
* Changed the Clutter::Actor::QUERY_COORDS() method signature: an ActorBox
  is also passed to the method, containing the current bounding box
  coordinates
* Added a new example showing how to correctly implementing an actor
* More documentation

Clutter 0.410
=============
* Fix Clutter::Actor::QUERY_COORDS to return an array
* Add a test for subclassing Clutter::Actor
* Documentation fixes in Clutter::Units and Clutter::Layout
* Allow a pure Perl implementation of Clutter::Layout::TUNE_REQUEST
* Bind more Clutter::Effect API
* Require the new stable release of Clutter

Clutter 0.402
=============
* Make subclassing Clutter::Box more perlish
* Add Clutter->SUPPORTED_MODULES(), to check the supported
  sub-modules against which Clutter was compiled
* Make Clutter::TestHelper skip tests if no support for the
  needed module was found

Clutter 0.401
=============
* Build fixes for packaging
* Add optional bindings for the Gtk2 integration
* Add a Test::More wrapper module and update the test suite

Clutter 0.400
=============
* Resync with upstream API
* Include autogenerated key symbols as Clutter::Keysyms
* Allow overriding Clutter::Actor virtual methods
* Add optional bindings for the GStreamer integration
* Add optional bindings for the Cairo integration
* Register a sink function for every initially unowned class
* Add more tests
Clutter 0.402
=============
* Make subclassing Clutter::Box more perlish

Clutter 0.401
=============
* Build fixes for packaging
* Add optional bindings for the Gtk2 integration
* Add a Test::More wrapper module and update the test suite

Clutter 0.400
=============
* Resync with upstream API
* Include autogenerated key symbols as Clutter::Keysyms
* Allow overriding Clutter::Actor virtual methods
* Add optional bindings for the GStreamer integration
* Add optional bindings for the Cairo integration
* Register a sink function for every initially unowned class
* Add more tests


Makefile.PL
BEGIN { require 5.008; }

use strict;
use warnings;

use ExtUtils::MakeMaker;
use File::Spec;
use Cwd;

# minimum required version of dependancies we need to build
our %build_reqs = (
    'perl-ExtUtils-Depends'   => '0.300',
    'perl-ExtUtils-PkgConfig' => '1.030',
    'perl-Glib'               => '1.220',
    'perl-Cairo'              => '1.000',
    'perl-Pango'              => '1.140',
    'clutter'                 => '1.0.0',
);

my %PREREQ_PM = (
    'ExtUtils::Depends'   => $build_reqs{'perl-ExtUtils-Depends'},
    'ExtUtils::PkgConfig' => $build_reqs{'perl-ExtUtils-PkgConfig'},
    'Glib'                => $build_reqs{'perl-Glib'},
    'Cairo'               => $build_reqs{'perl-Cairo'},
    'Pango'               => $build_reqs{'perl-Pango'},
);

# Writing a fake Makefile ensures that CPAN will pick up the correct
# dependencies and install them.
unless (eval "use ExtUtils::Depends '$build_reqs{'perl-ExtUtils-Depends'}';"
           . "use ExtUtils::PkgConfig '$build_reqs{'perl-ExtUtils-PkgConfig'}';"
           . "use Glib '$build_reqs{'perl-Glib'}';"
           . "use Cairo '$build_reqs{'perl-Cairo'}';"
           . "use Pango '$build_reqs{'perl-Pango'}';"
	   . "use Glib::CodeGen;"
	   . "use Glib::MakeHelper;"
           . "1") {
   warn <<"__EOW__";
Pre-requisites not found:
$@
Please install them manually.
__EOW__
   exit 0;
}

mkdir 'build', 0777;

# Base module
my %pkgcfg;
unless (eval { %pkgcfg = ExtUtils::PkgConfig->find('clutter-1.0'); 1; }) {
    warn $@;
    exit 0;
}

# retrieve Clutter's version
my @clutter_version = split /\./, $pkgcfg{modversion};

# decide what files to use in the build based on the version we found.
my @xs_files = ();
my @xs_lists = Glib::MakeHelper->select_files_by_version("xs_files", @clutter_version);
foreach my $filename (@xs_lists) {
    my @names = Glib::MakeHelper->read_source_list_file($filename);
    push @xs_files, @names;
}

print "Loaded " . scalar(@xs_files) . " xs files for Clutter\n";

my %pm_files = (
    'lib/Clutter.pm' => '$(INST_LIBDIR)/Clutter.pm',
    'Cookbook.pod' => '$(INST_LIBDIR)/Clutter/Cookbook.pod',

    # Keysyms.pm is generated by Makefile.PL
    'build/Keysyms.pm' => '$(INST_LIBDIR)/Clutter/Keysyms.pm',

    # TestHelper is just a wrapper around Test::More
    'tools/TestHelper.pm' => '$(INST_LIBDIR)/Clutter/TestHelper.pm',
);

my %pod_files = (
    'lib/Clutter.pm' => '$(INST_MAN3DIR)/Clutter.$(MAN3EXT)',
    'Cookbook.pod' => '$(INST_MAN3DIR)/Clutter::Cookbook.$(MAN3EXT)',

    # Keysyms.pm is generated by Makefile.PL
    'build/Keysyms.pm' => '$(INST_MAN3DIR)/Clutter::Keysyms.$(MAN3EXT)',
    
    Glib::MakeHelper->do_pod_files(@xs_files),
);

# typemaps generation files
my @input_maps = qw/ maps /;

# autogeneration
Glib::CodeGen->parse_maps('clutter-perl', input => [<maps-[0-9]\.[0-9]*>]);
Glib::CodeGen->write_boot(xs_files => \@xs_files);

# this one turns the long list of #defines in clutter-keysyms.h into a
# large hash of key names and values, %Clutter::Gdk::Keysyms.
system "$^X tools/genkeysyms.pl >\"build/Keysyms.pm\"";

my @deps = qw/ Glib Cairo Pango/;
my $cwd = cwd();

my $module_cflags = $pkgcfg{cflags};
my $module_libs   = $pkgcfg{libs};

my @typemaps = map { File::Spec->catfile($cwd, $_) }
               qw( build/clutter-perl.typemap clutter.typemap );

my $clutter = ExtUtils::Depends->new('Clutter', @deps);
$clutter->set_inc($module_cflags . ' -I./build ');
$clutter->set_libs($module_libs);
$clutter->add_xs(@xs_files);
$clutter->add_pm(%pm_files);
$clutter->add_typemaps(@typemaps);
$clutter->install(qw(clutter-perl.h build/clutter-perl-autogen.h doctypes));
$clutter->save_config('build/IFiles.pm');

WriteMakefile(
    AUTHOR        => 'Emmanuele Bassi <ebassi@linux.intel.com>',
    NAME          => 'Clutter',
    VERSION_FROM  => 'lib/Clutter.pm',
    ABSTRACT_FROM => 'lib/Clutter.pm',
    LICENSE       => 'perl, lgpl',

    # what the module requires to run
    PREREQ_PM     => \%PREREQ_PM,
    PREREQ_FATAL  => 1,

    XSPROTOARG    => '-noprototypes ',
    MAN3PODS	  => \%pod_files,
    DL_FUNCS      => { Clutter => [] },

    META_MERGE    => { configure_requires => \%PREREQ_PM },

    $clutter->get_makefile_vars(),
);


=pod
#=unstable

print "
WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING

This is an unstable development release of Clutter.  The API is
not frozen and things are subject to change at any time.  Report any
bugs to clutter AT openedhand DOT com as soon as possible.

WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING
";

=cut

sub MY::postamble
{
    return Glib::MakeHelper->postamble_clean()
         . Glib::MakeHelper->postamble_docs_full(
               DEPENDS => $clutter,
               DOCTYPES => 'doctypes',
               COPYRIGHT_FROM => 'copyright.pod',
           );
}
